### PR TITLE
 Escape HTML using the velocity engine

### PR DESCRIPTION
**Pull request by mpkorstanje**
It is preferable to escape HTML through the template engine over escaping it at in
the template. This ensures none can be forgotten on accident. The unfortunate side
effect of this is that it no longer possible to  provide html as an argument in a
macro. This leads to some duplication in embeddings.vm

 Existing methods to escape html have been removed.

### CODE REVIEW

**Comment by mpkorstanje** [comment]
First draft of a fundamental solution for  #556. 

Still a WIP but open for comments.

**Comment by mpkorstanje** [comment]
Visual inspection of sample.json report looks good.

File: pom.xml [hunk 1]
```diff
@@ -228,6 +228,13 @@
             <artifactId>velocity-tools</artifactId>
             <version>1.4</version>
         </dependency>
+
+        <dependency>
+            <groupId>oro</groupId>
+            <artifactId>oro</artifactId>
+            <version>2.0.8</version>
+        </dependency>
```
Comment (damianszczepanik) [inline pom.xml]: I was not able to run this code because of
``
ERROR StatusLogger No log4j2 configuration file found. Using default configuration: logging only errors to the console.
Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/oro/text/perl/MalformedPerl5PatternException
``

File: src/main/resources/templates/macros/json/embeddings.vm [hunk 2]
```diff
@@ -31,39 +31,63 @@
 
 
 #macro(includeImageEmbedding, $embedding, $image_type, $index)
-  #set($Quote = '"')
-  #includeExpandableEmbedding($embedding, $image_type, "<img src=${Quote}embeddings/$embedding.getFileName()${Quote}>", $index)
+  #set($index = $index + 1)
+  <div class="embedding indention">
+    #set($embeddingId = $counter.next())
```
Comment (damianszczepanik) [inline src/main/resources/templates/macros/json/embeddings.vm]: what is going on here, why the change is so big?

File: src/test/java/net/masterthought/cucumber/generators/integrations/FailuresOverviewPageIntegrationTest.java [hunk 3]
```diff
@@ -69,21 +69,4 @@ public void generatePage_onJsonWithoutFailedSteps_generatesProperMessage() {
         SummaryAssertion summary = document.getReport();
         assertThat(summary.getEmptyReportMessage()).isEqualTo("You have no failed scenarios in your Cucumber report");
     }
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/integrations/FailuresOverviewPageIntegrationTest.java]: why those tests were removed

File: src/main/java/net/masterthought/cucumber/generators/AbstractPage.java [hunk 4]
```diff
@@ -79,6 +80,7 @@ private Properties buildProperties() {
         props.setProperty("resource.loader", "class");
         props.setProperty("class.resource.loader.class", ClasspathResourceLoader.class.getCanonicalName());
         props.setProperty("runtime.log", new File(configuration.getReportDirectory(), "velocity.log").getPath());
+        props.setProperty("eventhandler.referenceinsertion.class", "org.apache.velocity.app.event.implement.EscapeHtmlReference");
```
Comment (damianszczepanik) [inline src/main/java/net/masterthought/cucumber/generators/AbstractPage.java]: so by this all labels are escaped so you are not able to provide HTML code into the report and with it will be displayed as HTML (eg link to external server in description)?

File: pom.xml (hunk 1 above)
Comment (mpkorstanje) [inline pom.xml]: You may need to do a `mvn clean install` first or if you're using an IDE reload or refresh the project. Looks like it is not downloading the dependency.

Comment (damianszczepanik) [inline pom.xml]: mvn clean test works fine so dependencies are fine

File: src/test/java/net/masterthought/cucumber/generators/integrations/FailuresOverviewPageIntegrationTest.java (hunk 3 above)
Comment (mpkorstanje) [inline src/test/java/net/masterthought/cucumber/generators/integrations/FailuresOverviewPageIntegrationTest.java]: They were testing methods that escaped data. These methods were no longer used and thus removed, leaving nothing to be tested.

File: src/main/resources/templates/macros/json/embeddings.vm (hunk 2 above)
Comment (mpkorstanje) [inline src/main/resources/templates/macros/json/embeddings.vm]: The `includeExpandableEmbedding` macro was used like this:

    #includeExpandableEmbedding($embedding, $format,
                <pre>$embedding.getEscapedDecodedData()</pre>", $index) 

Escaping all html broke the  macro because the value of `$content` would also be escaped. I fixed this by inlining `includeExpandableEmbedding` into every macro that used it. This results in the heavy duplication that you're looking at now.

File: src/main/java/net/masterthought/cucumber/generators/AbstractPage.java (hunk 4 above)
Comment (mpkorstanje) [inline src/main/java/net/masterthought/cucumber/generators/AbstractPage.java]: The way it is currently setup that is correct. 

However it is possible to exclude some labels from being escaped. Though I would not recommend this. Rather I would suggest adding a markdown parser and use markdown formatting to provide the markup.

You can read about the functions and options that velocity provides here:

https://velocity.apache.org/engine/devel/developer-guide.html#orgapachevelocityappeventreferenceinsertioneventhandler

File: pom.xml (hunk 1 above)
Comment (mpkorstanje) [inline pom.xml]: Interesting. You should not be able to pass the tests without oro. EscapeReference the parent of EscapeHtmlReference uses oro. Can you provide a simple reproducer?

File: src/main/java/net/masterthought/cucumber/generators/AbstractPage.java (hunk 4 above)
Comment (mpkorstanje) [inline src/main/java/net/masterthought/cucumber/generators/AbstractPage.java]: Having looked at [flexmark](https://github.com/vsch/flexmark-java)  I think it would be fairly simple to add. Though I also think that should be done on another pull request.

Comment (mpkorstanje) [inline src/main/java/net/masterthought/cucumber/generators/AbstractPage.java]: But I suppose that allowing some html elements in the feature section is permissible.

```
<a>              - hyperlink.
<b>              - bold, use as last resort <h1>-<h3>, <em>, and <strong> are 
                   preferred.
<blockquote>     - specifies a section that is quoted from another source.
<code>           - defines a piece of computer code.
<del>            - delete, used to indicate modifications.
<dd>             - describes the item in a <dl> description list.
<dl>             - description list.
<dt>             - title of an item in a <dl> description list.
<em>             - emphasized.
<h1>, <h2>, <h3> - headings.
<i>              - italic.
<img>            - specifies an image tag.
<kbd>            - represents user input (usually keyboard input).
<li>             - list item in an ordered list <ol> or an unordered list <ul>.
<ol>             - ordered list.
<p>              - paragraph.
<pre>            - pre-element displayed in a fixed width font and and 
                   unchanged line breaks.
<s>              - strikethrough.
<sup>            - superscript text appears 1/2 character above the baseline 
                   used for footnotes and other formatting.
<sub>            - subscript appears 1/2 character below the baseline.
<strong>         - defines important text.
<strike>         - strikethrough is deprecated, use <del> instead.
<ul>             - unordered list.
<br>             - line break.
<hr>             - defines a thematic change in the content, usually via a 
                   horizontal line.
```

File: pom.xml (hunk 1 above)
Comment (damianszczepanik) [inline pom.xml]: https://github.com/damianszczepanik/cucumber-reporting#usage here you have some code how to run the report as the standalone app - I use this code and then the code fails.

You can use also this real code from the Jenkins plugin https://github.com/jenkinsci/cucumber-reports-plugin/blob/master/src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java

File: src/main/java/net/masterthought/cucumber/generators/AbstractPage.java (hunk 4 above)
Comment (damianszczepanik) [inline src/main/java/net/masterthought/cucumber/generators/AbstractPage.java]: I'm not sure If I want to display doc_string should be displayed as escaped html but let's go for this and if there will be other requests I can handle this

File: src/test/java/net/masterthought/cucumber/generators/integrations/FailuresOverviewPageIntegrationTest.java (hunk 3 above)
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/integrations/FailuresOverviewPageIntegrationTest.java]: But then we have no test that validates if data is escaped properly (which is not global. default configuration). Worth to have test for this, even one.

File: src/main/java/net/masterthought/cucumber/generators/AbstractPage.java (hunk 4 above)
Comment (damianszczepanik) [inline src/main/java/net/masterthought/cucumber/generators/AbstractPage.java]: Also I prefer compile time checking, not runtime checking
EscapeHtmlReference.class.getName() should work

Comment (mpkorstanje) [inline src/main/java/net/masterthought/cucumber/generators/AbstractPage.java]: I'm sorry. I can't quite parse your intentions.

Should the report look like this

> 
> Scenario: A scenario
> 
> <i>Some text here with safe html markup.</i>
> 
> Feature: Docstring
>    Given a doc string containing html mark up
>    """
>   \<i\>Some text here with html markup\</i\>
>    """
> 

Or like this:

> 
> Scenario: A scenario
> 
> \<i\>Some text here with safe html markup.\</i\>
> 
> Feature: Docstring
>    Given a doc string containing html mark up
>    """
>   \<i\>Some text here with html markup\</i\>
>    """
>

File: src/test/java/net/masterthought/cucumber/generators/integrations/FailuresOverviewPageIntegrationTest.java (hunk 3 above)
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/integrations/FailuresOverviewPageIntegrationTest.java]: let's leave separate

File: src/main/java/net/masterthought/cucumber/generators/AbstractPage.java (hunk 4 above)
Comment (damianszczepanik) [inline src/main/java/net/masterthought/cucumber/generators/AbstractPage.java]: First one as I don't want to change data from the json file. Whatever you pass as doc_string it will be displayed as it is. This allows you to have html code in generated report. If you want to escape html into plain text maybe you should do it by yourself before report is started?

I have doubts for this so at this point all cases are fine as long as more users expect to have doc_string escaped let's leave version you prepared. In case of problem we will investigate and prepare new PR

Comment (mpkorstanje) [inline src/main/java/net/masterthought/cucumber/generators/AbstractPage.java]: Mmh. I don't understand why you call it 'unchanged'

Suppose I have a json file that contains:

    <b> a bold statement </b>

When rendered by a plain text reporter, like the command-line, it will be rendered as 

    <b> a bold statement </b>

And when rendered by a html reporter that doesn't escape the input
<b> a bold statement </b>
Because now I've lost the \<b\> \</b\> tags from the report. 

---

I also don't understand why you expect the html to be escaped in the json file if it should be displayed escaped. When creating the json file and escaping html I've coupled that json file to a particular report formatter (one that uses html). This seems bad practice.

File: pom.xml [hunk 5]
```diff
@@ -228,6 +228,7 @@
             <artifactId>velocity-tools</artifactId>
             <version>1.4</version>
         </dependency>
+
```
Comment (damianszczepanik) [inline pom.xml]: revert the whole file if nothing has changed

File: src/main/java/net/masterthought/cucumber/generators/AbstractPage.java [hunk 6]
```diff
@@ -84,6 +88,11 @@ private Properties buildProperties() {
     }
 
     private void buildGeneralParameters() {
+        // to escape html
```
Comment (damianszczepanik) [inline src/main/java/net/masterthought/cucumber/generators/AbstractPage.java]: html and xml :)

File: src/main/java/net/masterthought/cucumber/generators/AbstractPage.java [hunk 7]
```diff
@@ -107,4 +116,20 @@ private void buildGeneralParameters() {
...
     }
+
+    /**
+     * Escapes all html that was provided in a reference before inserting it into a template.
+     */
+    private final static class EscapeHtmlReference implements ReferenceInsertionEventHandler {
+
+        @Override
+        public Object referenceInsert(String reference, Object value) {
+            if (value == null) {
```
Comment (damianszczepanik) [inline src/main/java/net/masterthought/cucumber/generators/AbstractPage.java]: when the value can be null?

File: src/main/java/net/masterthought/cucumber/json/Result.java [hunk 8]
```diff
@@ -27,8 +25,7 @@ public String getFormatedDuration() {
         return Util.formatDuration(duration);
     }
 
-    /** Returns error message escaped so it is properly displayed as HTML. */
     public String getErrorMessage() {
-        return StringEscapeUtils.escapeHtml(errorMessage);
+        return errorMessage;
```
Comment (damianszczepanik) [inline src/main/java/net/masterthought/cucumber/json/Result.java]: did you remove 'escape' operation from all methods?

File: src/main/resources/templates/macros/json/embeddings.vm [hunk 9]
```diff
@@ -10,7 +10,7 @@
     #elseif ($embedding.getMimeType() == "image/jpeg")
       #includeImageEmbedding($embedding, "jpeg", $foreach.index)
     #elseif ($embedding.getMimeType() == "text/xml")
-      #includeEscapedTextEmbedding($embedding, "XML text", $foreach.index)
+      #includeTextEmbedding($embedding, "XML text", $foreach.index)
```
Comment (damianszczepanik) [inline src/main/resources/templates/macros/json/embeddings.vm]: just to be sure 'includeEscapedTextEmbedding' was removed from the code?

File: src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java [hunk 10]
```diff
@@ -48,8 +47,32 @@ public void generateReport_CreatesReportFile() {
         assertThat(reportFile).exists();
     }
 
+
+    @Test
+    public void generateReport_EscapesHtml() {
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java]: generateReport_DisplaysContentAsEscapedText

File: src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java [hunk 11]
```diff
@@ -48,8 +47,32 @@ public void generateReport_CreatesReportFile() {
...
+                equalTo("<b>a bold statement</b>"));
+        assertThat(document.getFeature().getElements()[0].getStepsSection().getSteps()[0].getDocString().text(),
+                equalTo("Doc string <b>a bold\nstatement</b>"));
+        assertThat(document.getFeature().getElements()[0].getStepsSection().getSteps()[1].getArgumentsTable().getBodyRow().getCellsValues()[0],
+                equalTo("<b>bold</b>"));
+    }
+
     @Test
     public void generateReport_OnInvalidPath_ThrowsException() {
+        setUpWithJson(SAMPLE_JSON);
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java]: why it was not needed and now you copy it over and over again?

**Comment by mpkorstanje** [comment]
I've removed the dependency on Oro. Does it work for you now?

File: src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java [hunk 12]
```diff
@@ -1,11 +1,14 @@
 package net.masterthought.cucumber.generators;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.hamcrest.CoreMatchers.equalTo;
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java]: we don't use this equal method, compare with other tests

File: src/test/resources/json/html.json [hunk 13]
```diff
@@ -0,0 +1,71 @@
+[
```
Comment (damianszczepanik) [inline src/test/resources/json/html.json]: don't need to have separate JSON if you use current sample JSON or test via unit test

Comment (mpkorstanje) [inline src/test/resources/json/html.json]: The current sample json doesn't contain enough html.

I'd also rather not modify it as FeaturesOverviewPageIntegrationTest#generatePage_generatesStatsTableFooter and some other tests rely on statistics of that exact file.

File: src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java (hunk 11 above)
Comment (mpkorstanje) [inline src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java]: Because leaving it in the @Before method and then calling setUpWithJson(HTML_JSON); in the escapeHtml test would be deceptive.

File: src/main/resources/templates/macros/json/embeddings.vm (hunk 9 above)
Comment (mpkorstanje) [inline src/main/resources/templates/macros/json/embeddings.vm]: Yes. There is no need for it any more.

File: src/main/java/net/masterthought/cucumber/generators/AbstractPage.java (hunk 7 above)
Comment (mpkorstanje) [inline src/main/java/net/masterthought/cucumber/generators/AbstractPage.java]: There is no specification so I copied that detail from the EscapeReference implementation but if you follow the data flow up to ASTReference.render you'll see that null is a plausible value.

Presumably when using

```
 $object.methodThatReturnsNull()
```

File: src/main/java/net/masterthought/cucumber/json/Result.java (hunk 8 above)
Comment (mpkorstanje) [inline src/main/java/net/masterthought/cucumber/json/Result.java]: Except in a few tests that were now comparing unescaped text against escaped text.

**Comment by mpkorstanje** [comment]
Commits have been squashed and force pushed. They're in this PR now.

File: src/main/java/net/masterthought/cucumber/generators/AbstractPage.java (hunk 7 above)
Comment (damianszczepanik) [inline src/main/java/net/masterthought/cucumber/generators/AbstractPage.java]: OK

File: src/main/java/net/masterthought/cucumber/generators/AbstractPage.java [hunk 14]
```diff
@@ -107,4 +115,20 @@ private void buildGeneralParameters() {
             }
         }
     }
+
+    /**
+     * Escapes all html and xml that was provided in a reference before inserting it into a template.
+     */
+    private final static class EscapeHtmlReference implements ReferenceInsertionEventHandler {
```
Comment (damianszczepanik) [inline src/main/java/net/masterthought/cucumber/generators/AbstractPage.java]: As you see there are many tests (not only unit) that reduces number of regressions when preparing new PR.
Can you add unit test for both cases?

File: src/main/resources/templates/macros/json/embeddings.vm [hunk 15]
```diff
@@ -31,39 +31,63 @@
...
-
-#macro(includeExpandableEmbedding, $embedding, $type, $content, $index)
   #set($index = $index + 1)
   <div class="embedding indention">
-    #set($embeddingId = $counter.next())
+  #set($embeddingId = $counter.next())
     <div data-toggle="collapse" data-target="#embedding-$embeddingId" class="collapsable-control">
-      <a>Attachment $index ($type)</a>
-      <a href="embeddings/$embedding.getFileName()" target="_blank"><span class="download-button glyphicon glyphicon-download-alt"></span></a>
+        <a>Attachment $index ($embedding.getMimeType())</a>
```
Comment (damianszczepanik) [inline src/main/resources/templates/macros/json/embeddings.vm]: I use two spaces indention, if you don't mind

File: src/test/java/net/masterthought/cucumber/generators/integrations/FailuresOverviewPageIntegrationTest.java [hunk 16]
```diff
@@ -69,21 +69,4 @@ public void generatePage_onJsonWithoutFailedSteps_generatesProperMessage() {
...
-
-    @Test
-    public void generatePage_generatesSummary() {
-
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/integrations/FailuresOverviewPageIntegrationTest.java]: this test checks if for the 'failure overview' page there is only one scenario with expected name - removing this test we loose validation what is really displayed on the page

File: src/test/java/net/masterthought/cucumber/json/FeatureTest.java [hunk 17]
```diff
@@ -46,20 +46,6 @@ public void getId_ReturnsID() {
     }
 
     @Test
-    public void getElements_ReturnsElements() {
-
-        // given
-        Feature feature = features.get(0);
-
-        // when
-        Element[] elements = feature.getElements();
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/json/FeatureTest.java]: do we have unit test for getElements() now?

File: src/test/resources/json/html.json (hunk 13 above)
Comment (damianszczepanik) [inline src/test/resources/json/html.json]: I know this is not the prefect solution as it is sometimes pain to maintain but as long as there are so many tests - tests says which steps should be updated and it's easy

I prefer to have one, complex json file with all options than having many files. Add required items to existing file (only required items

Comment (damianszczepanik) [inline src/test/resources/json/html.json]: BTW, you can rebase code

App is working correctly now (when you removed lib dependency)

File: src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java [hunk 18]
```diff
@@ -48,8 +43,32 @@ public void generateReport_CreatesReportFile() {
...
+
+        // when
+        page.generatePage();
+
+        // then
+        DocumentAssertion document = documentFrom(page.getWebPage());
+        assertThat(document.getFeature().getDescription())
+                .isEqualTo("<b>a bold statement</b>");
+        assertThat(document.getFeature().getElements()[0].getDescription())
+                .isEqualTo("<b>a bold statement</b>");
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java]: name of the method says that data should be escaped while tests validates against not-escaped content - can you tell me more about it?

File: src/main/resources/templates/macros/json/embeddings.vm [hunk 19]
```diff
@@ -31,39 +31,63 @@
...
+        <a href="embeddings/$embedding.getFileName()" target="_blank"><span class="download-button glyphicon glyphicon-download-alt"></span></a>
     </div>
     <div id="embedding-$embeddingId" class="collapse collapsable-details">
-      <span class="embedding-content">$content</span>
+        <span class="embedding-content">This file cannot be displayed. Use download button to get the content as file.</span>
     </div>
   </div>
-#end
+#end
```
Comment (damianszczepanik) [inline src/main/resources/templates/macros/json/embeddings.vm]: empty line is missing

File: src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java [hunk 20]
```diff
@@ -1,24 +1,23 @@
...
 import net.masterthought.cucumber.util.Counter;
 import net.masterthought.cucumber.util.Util;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.velocity.VelocityContext;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+import java.io.File;
+import java.util.Properties;
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java]: usually java package is located the most top, but anyway

File: src/main/java/net/masterthought/cucumber/generators/AbstractPage.java [hunk 21]
```diff
@@ -84,6 +87,11 @@ private Properties buildProperties() {
     }
 
     private void buildGeneralParameters() {
+        // to escape html and xml
+        EventCartridge ec = new EventCartridge();
```
Comment (mpkorstanje) [inline src/main/java/net/masterthought/cucumber/generators/AbstractPage.java]: That would put the EventCartridge and EscapeHtmlReference in a scope that is greater then strictly needed. Neither is expensive to create nor created often so I don't see a good performance reason either.

File: src/main/java/net/masterthought/cucumber/generators/AbstractPage.java (hunk 14 above)
Comment (mpkorstanje) [inline src/main/java/net/masterthought/cucumber/generators/AbstractPage.java]: Ok.

File: src/main/resources/templates/macros/json/embeddings.vm (hunk 15 above)
Comment (mpkorstanje) [inline src/main/resources/templates/macros/json/embeddings.vm]: Np. Just a formatter setting.

File: src/test/java/net/masterthought/cucumber/generators/integrations/FailuresOverviewPageIntegrationTest.java (hunk 16 above)
Comment (mpkorstanje) [inline src/test/java/net/masterthought/cucumber/generators/integrations/FailuresOverviewPageIntegrationTest.java]: Check.

File: src/test/java/net/masterthought/cucumber/json/FeatureTest.java (hunk 17 above)
Comment (mpkorstanje) [inline src/test/java/net/masterthought/cucumber/json/FeatureTest.java]: Check.

File: src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java (hunk 20 above)
Comment (mpkorstanje) [inline src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java]: Ok. Will rebase on next PR.

The package order come from the default settings. In .idea. I'll see if I can find that setting.

File: src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java (hunk 18 above)
Comment (mpkorstanje) [inline src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java]: If you drill down deep enough into `getDescription` you'll see that it uses `TextNode.getWholeText()` which uses the [html text attribute](http://www.w3schools.com/jsref/prop_node_textcontent.asp). This returns the textual representation of the nodes content. Because it is the textual representation any escaped characters are no longer escaped. As it should be, because this is what the user sees.

File: src/test/java/net/masterthought/cucumber/generators/integrations/helpers/EmbeddingAssertion.java [hunk 22]
```diff
@@ -19,7 +19,8 @@ public void hasImageContent(Embedding embedding) {
     }
 
     public void hasTextContent(String content) {
-        assertThat(getBox().oneBySelector("pre", WebAssertion.class).html()).isEqualTo(getDecodedData(content));
+        assertThat(getBox().oneBySelector("pre", WebAssertion.class).text())
+                .isEqualTo(getDecodedData(content));
     }
```
Comment (mpkorstanje) [inline src/test/java/net/masterthought/cucumber/generators/integrations/helpers/EmbeddingAssertion.java]: Replaced .html with .text so the decoded data doesn't need to be escaped to match up.

File: src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java [hunk 23]
```diff
@@ -48,6 +49,33 @@ public void generateReport_CreatesReportFile() {
...
+                .isEqualTo("Error message java.lang.AssertionError: \n" +
+                        "Expected: is <80>\n" +
+                        "     got: <90>\n" +
+                        "\n" +
+                        "\tat org.junit.Assert.assertThat(Assert.java:780)\n" +
+                        "\tat org.junit.Assert.assertThat(Assert.java:738)\n" +
+                        "\tat net.masterthought.example.ATMScenario.checkBalance(ATMScenario.java:69)\n" +
+                        "\tat ✽.And the account balance should be 90(net/masterthought/example/ATMK.feature:12)");
+    }
+
```
Comment (mpkorstanje) [inline src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java]: Best I can make of it. Works for the purpose of verifying that html is escaped.

**Comment by mpkorstanje** [comment]
Rebased and edits squashed.

File: src/test/java/net/masterthought/cucumber/generators/EscapeHtmlReferenceTest.java [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+ */
+public class EscapeHtmlReferenceTest {
+
+    private static final String SOME_REFERENCE = "someReference";
+    private final ReferenceInsertionEventHandler insertionEventHandler = new EscapeHtmlReference();
+
+    @Test
+    public void referenceInsert_returnNormalText(){
+        String normalText = "a plain statement";
+        assertThat(insertionEventHandler.referenceInsert(SOME_REFERENCE, normalText)).isEqualTo(normalText);
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/EscapeHtmlReferenceTest.java]: I prefer https://en.wikipedia.org/wiki/Behavior-driven_development as the way how tests are implemented

File: src/test/java/net/masterthought/cucumber/generators/EscapeHtmlReferenceTest.java [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+
+    @Test
+    public void referenceInsert_shouldEscapeHtmlForAnyLabel(){
+        String html = "<b>a bold statement</b>";
+        assertThat(insertionEventHandler.referenceInsert(SOME_REFERENCE, html)).isEqualTo(escapeHtml(html));
+    }
+
+    @Test
+    public void referenceInsert_shouldReturnNullForNull(){
+        assertThat(insertionEventHandler.referenceInsert(SOME_REFERENCE, null)).isNull();
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/EscapeHtmlReferenceTest.java]: I'm happy you covered all cases

File: src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java [hunk 26]
```diff
@@ -48,6 +49,33 @@ public void generateReport_CreatesReportFile() {
...
+        // when
+        page.generatePage();
+
+        // then
+        DocumentAssertion document = documentFrom(page.getWebPage());
+        assertThat(document.getFeature().getDescription())
+                .isEqualTo("As an Account Holder I want to withdraw cash from an ATM,<br>so that I can get money when the bank is closed");
+        assertThat(document.getFeature().getElements()[0].getStepsSection().getSteps()[5].getEmbedding()[3].text())
+                .isEqualTo("Attachment 4 (HTML) <i>Hello</i> <b>World!</b>");
+        assertThat(document.getFeature().getElements()[0].getStepsSection().getSteps()[5].getMessage().text())
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java]: text() could be hidden inside of getMessage()

**Comment by damianszczepanik** [comment]
Good job

**Comment by damianszczepanik** [comment]
@mpkorstanje

There is regression - charts for tags and trends are not displayed

**Comment by mpkorstanje** [comment]
Taking a look. Probely missed some more html injection.

**Comment by mpkorstanje** [comment]
Fixed with #584.
